authz/providers/azure/rbac/rbac.go

```diff
@@ -0,0 +1,283 @@
+	rbacURL, err := url.Parse(authzInfo.ARMEndPoint)
+	if err != nil {
+		return nil, err
+	}
+
+	var tokenProvider graph.TokenProvider
+	switch opts.AuthzMode {
+	case authzOpts.ARCAuthzMode:
+		tokenProvider = graph.NewClientCredentialTokenProvider(authopts.ClientID, authopts.ClientSecret,
+			fmt.Sprintf("%s%s/oauth2/v2.0/token", authzInfo.AADEndpoint, authopts.TenantID),
+			fmt.Sprintf("%s.default", authzInfo.ARMEndPoint))
+	case authzOpts.AKSAuthzMode:
+		tokenProvider = graph.NewAKSTokenProvider(opts.AKSAuthzTokenURL, authopts.TenantID)
+	}
+
+	return newAccessInfo(tokenProvider, rbacURL, opts)
+}
+
+func (a *AccessInfo) RefreshToken() error {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+	if a.IsTokenExpired() {
+		resp, err := a.tokenProvider.Acquire("")
+		if err != nil {
+			glog.Errorf("%s failed to refresh token : %s", a.tokenProvider.Name(), err.Error())
+			return errors.Wrap(err, "failed to refresh rbac token")
+		}
+
+		// Set the authorization headers for future requests
+		a.headers.Set("Authorization", fmt.Sprintf("Bearer %s", resp.Token))
+		expIn := time.Duration(resp.Expires) * time.Second
+		a.expiresAt = time.Now().Add(expIn - expiryDelta)
+		glog.Infof("Token refreshed successfully on %s. Expire at:%s", time.Now(), a.expiresAt)
+	}
+
+	return nil
+}
+
+func (a *AccessInfo) IsTokenExpired() bool {
+	return a.expiresAt.Before(time.Now())
+}
+
+func (a *AccessInfo) ShouldSkipAuthzCheckForNonAADUsers() bool {
+	return a.skipAuthzForNonAADUsers
+}
+
+func (a *AccessInfo) GetResultFromCache(request *authzv1.SubjectAccessReviewSpec, store authz.Store) (bool, bool) {
+	var result bool
+	key := getResultCacheKey(request)
+	glog.V(10).Infof("Cache search for key: %s", key)
+	found, _ := store.Get(key, &result)
+	return found, result
+}
+
+func (a *AccessInfo) SkipAuthzCheck(request *authzv1.SubjectAccessReviewSpec) bool {
+	if a.clusterType == connectedClusters {
+		_, ok := a.skipCheck[strings.ToLower(request.User)]
+		return ok
+	}
+	return false
+}
+
+func (a *AccessInfo) SetResultInCache(request *authzv1.SubjectAccessReviewSpec, result bool, store authz.Store) error {
+	key := getResultCacheKey(request)
+	glog.V(10).Infof("Cache set for key: %s, value: %t", key, result)
+	return store.Set(key, result)
+}
+
+func (a *AccessInfo) AllowNonResPathDiscoveryAccess(request *authzv1.SubjectAccessReviewSpec) bool {
+	if request.NonResourceAttributes != nil && a.allowNonResDiscoveryPathAccess && strings.EqualFold(request.NonResourceAttributes.Verb, "get") {
+		path := strings.ToLower(request.NonResourceAttributes.Path)
+		if strings.HasPrefix(path, "/api") || strings.HasPrefix(path, "/openapi") || strings.HasPrefix(path, "/version") || strings.HasPrefix(path, "/healthz") {
+			return true
+		}
+	}
+	return false
+}
+
+func (a *AccessInfo) setReqHeaders(req *http.Request) {
+	a.lock.RLock()
+	defer a.lock.RUnlock()
+	// Set the auth headers for the request
+	req.Header = a.headers
```

Comment:
I am not sure this is enough. In Go, maps are pointer types. So, req.Header is still pointing to the same map. I think you need to make a copy.

```
if req.Header == nil {
	req.Header = make(http.Header)
}
for k, values := range a.headers {
  ...
}

```